### PR TITLE
dkregistry: update to latest hyper and hyper-rustls

### PR DESCRIPTION
**Pull request by lucab**

### CODE REVIEW

**Comment by lucab** [comment]
@steveeJ this should be ready to land. I invited you to @camallo/dkregistry-admins, so you should be able to review and merge.

File: src/v2/blobs.rs [hunk 1]
```diff
@@ -10,75 +10,78 @@ impl Client {
...
+                    Some(loc) => {
+                        trace!("Got Location header {:?}", loc);
+                        String::from_utf8(loc.as_bytes().to_vec()).ok()
                     }
                 };
                 if let Some(u) = redirect {
                     let new_url = match hyper::Uri::from_str(u.as_str()) {
                         Ok(u) => u,
-                        Err(e) => return Either::A(future::result(Err(hyper::Error::Uri(e)))),
+                        _ => return Either::A(future::err(Error::from("get_blob: wrong URL"))),
```
Comment (steveeJ) [inline src/v2/blobs.rs]: It might be useful to include the URL in the error message.

File: src/v2/config.rs [hunk 2]
```diff
@@ -16,8 +17,7 @@ impl Config {
     /// Initialize `Config` with default values.
     pub fn default(handle: &reactor::Handle) -> Self {
         Self {
-            config: hyper::client::Client::configure()
-                .connector(hyper_rustls::HttpsConnector::new(4, handle)),
+            config: hyper::client::Client::builder().build(hyper_rustls::HttpsConnector::new(4)),
```
Comment (steveeJ) [inline src/v2/config.rs]: That's the magic 4 here?

File: src/v2/manifest/mod.rs [hunk 3]
```diff
@@ -26,29 +25,32 @@ impl Client {
...
         )));
         let req = {
-            let accept_types = Accept(vec![mediatypes::MediaTypes::ManifestV2S2.to_qitem()]);
-            let mut r = self.new_request(hyper::Method::Get, url.clone());
-            r.headers_mut().set(accept_types);
+            let mut r = self.new_request(hyper::Method::GET, url.clone());
+            let mtype = mediatypes::MediaTypes::ManifestV2S2.to_string();
+            r.headers_mut().append(
+                header::ACCEPT,
+                header::HeaderValue::from_str(&mtype).unwrap(),
```
Comment (steveeJ) [inline src/v2/manifest/mod.rs]: `?` instead of `unwrap()`?

File: src/v2/manifest/mod.rs [hunk 4]
```diff
@@ -26,29 +25,32 @@ impl Client {
...
+            );
             r
         };
         let freq = self.hclient.request(req);
         let fres = freq
-            .map(move |r| {
+            .from_err()
+            .inspect(move |_| {
```
Comment (steveeJ) [inline src/v2/manifest/mod.rs]: What is being moved here?

File: src/v2/manifest/mod.rs [hunk 5]
```diff
@@ -71,48 +73,53 @@ impl Client {
...
-            }).from_err();
-        return Ok(Box::new(fres));
+            });
+        Ok(Box::new(fres))
     }
 }
 
-fn to_mimes(v: &[&str]) -> Result<Vec<QualityItem<mime::Mime>>> {
-    let res: Vec<QualityItem<mime::Mime>>;
+fn to_mimes(v: &[&str]) -> Result<Vec<String>> {
```
Comment (steveeJ) [inline src/v2/manifest/mod.rs]: Is there no `Mime` type available anymore? I'd prefer to keep it

File: src/v2/mod.rs [hunk 6]
```diff
@@ -61,38 +61,46 @@ pub use self::blobs::FutureBlob;
...
     pub fn configure(handle: &reactor::Handle) -> Config {
         Config::default(handle)
     }
 
-    fn new_request(&self, method: hyper::Method, url: hyper::Uri) -> hyper::client::Request {
-        let mut req = client::Request::new(method, url);
-        let host = hyper::header::Host::new(self.index.clone(), None);
-        req.headers_mut().set(host);
+    fn new_request(&self, method: hyper::Method, url: hyper::Uri) -> hyper::Request<hyper::Body> {
+        // TODO(lucab): get rid of all unwraps here.
```
Comment (steveeJ) [inline src/v2/mod.rs]: Yes please :-D

Comment (steveeJ) [reply src/v2/mod.rs]: But we can leave it in and follow-up with an issue to remove all TODOs

Comment (steveeJ) [reply src/v2/mod.rs]: And by removing I mean of course just to delete the TODO lines

File: src/v2/tags.rs [hunk 7]
```diff
@@ -20,32 +21,31 @@ impl Client {
...
                 };
-                let ok = match r.headers().get_raw("Content-type") {
+                let ok = match r.headers().get(header::CONTENT_TYPE) {
                     None => false,
-                    Some(ct) => ct.iter().any(|v| v == b"application/json"),
+                    Some(ct) => ct.to_str()?.starts_with("application/json"),
                 };
                 if !ok {
-                    return Err(hyper::Error::Header);
+                    return Err(Error::from("get_tags: wrong HTTP header"));
```
Comment (steveeJ) [inline src/v2/tags.rs]: is it viable to include the header content here?

File: src/v2/tags.rs [hunk 8]
```diff
@@ -20,32 +21,31 @@ impl Client {
...
                 }
                 Ok(r)
             }).and_then(|r| {
-                r.body().fold(Vec::new(), |mut v, chunk| {
-                    v.extend(&chunk[..]);
-                    futures::future::ok::<_, hyper::Error>(v)
-                })
+                r.into_body()
+                    .concat2()
+                    .map_err(|e| format!("get_tags: failed to fetch whole body: {}", e).into())
```
Comment (steveeJ) [inline src/v2/tags.rs]: .. *the* whole body ..

File: src/v2/manifest/mod.rs (hunk 4 above)
Comment (lucab) [reply src/v2/manifest/mod.rs]: `url`, for trace-logging.

File: src/v2/config.rs (hunk 2 above)
Comment (lucab) [reply src/v2/config.rs]: Number of DNS threads internally spawned by the TLS connector. I'll make this an explicit variable with a proper name.

**Comment by lucab** [comment]
@steveeJ thanks for the review, fixed and rebased.

**Comment by lucab** [comment]
@steveeJ I've added dependabot over here to try to avoid stale dependencies in the future.

**Pull request merged**